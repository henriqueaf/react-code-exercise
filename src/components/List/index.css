.loading-container {
  text-align: center;
}

.loading {
  -webkit-animation: fadeBlink 2s infinite;
  -moz-animation: fadeBlink 2s infinite;
  animation: fadeBlink 2s infinite;
}

@-moz-keyframes fadeBlink {
  0%   { opacity:1; }
  50%  { opacity:0; }
  100% { opacity:1; }
}
@-webkit-keyframes fadeBlink {
  0%   { opacity:1; }
  50%  { opacity:0; }
  100% { opacity:1; }
}
@keyframes fadeBlink {
  0%   { opacity:1; }
  50%  { opacity:0; }
  100% { opacity:1; }
}
