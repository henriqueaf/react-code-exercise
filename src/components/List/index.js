import React, { useEffect, useState } from 'react';
import { Button } from 'react-bootstrap';
import includes from 'lodash/includes';
import toLower from 'lodash/toLower';
import replace from 'lodash/replace';
import './index.css';

import Filter from '../Filter';
import Table from '../Table';
import ListItem from '../ListItem';
import Pagination from '../Pagination';
import CongressImg from '../../images/congress.png';

const HOUSE_MINIMUM_SESSION = 102;
const SENATE_MINIMUM_SESSION = 80;

export default () => {
  const [loading, setLoading] = useState(false);
  const [members, setMembers] = useState([]);
  const [filteredMembers, setFilteredMembers] = useState([]);

  const [membersPerPage, setMembersPerPage] = useState(5);
  const [currentPage, setCurrentPage] = useState(0);
  const [chamber, setChamber] = useState('senate');
  const [session, setSession] = useState(115);
  const [minimumSession, setMinimumSession] = useState(SENATE_MINIMUM_SESSION);

  const startIndex = currentPage * membersPerPage;
  const endIndex = startIndex + membersPerPage;

  const calculateMinimumSession = () => {
    switch (chamber) {
      case 'senate':
        return SENATE_MINIMUM_SESSION;
        break;
      case 'house':
        return HOUSE_MINIMUM_SESSION;
        break;
    }
  };

  useEffect(() => {
    setLoading(true);

    fetch(`https://api.propublica.org/congress/v1/${session}/${chamber}/members.json`, {
      headers: new Headers({
        'X-API-Key': 'd0ywBucVrXRlMQhENZxRtL3O7NPgtou2mwnLARTr',
      }),
    })
      .then((res) => res.json())
      .then((json) => json.results[0].members)
      .then((responseMembers) => {
        console.log(responseMembers);
        setMembers(responseMembers);
        setFilteredMembers(responseMembers);
      })
      .catch(() => {
        setLoading(false);
      })
      .finally(() => {
        setLoading(false);
        setMinimumSession(calculateMinimumSession());
      });
  }, [chamber, session]);

  useEffect(() => {
    setCurrentPage(0);
  }, [membersPerPage, filteredMembers]);

  useEffect(() => {
    if (chamber === 'house' && session < HOUSE_MINIMUM_SESSION)
      setSession(HOUSE_MINIMUM_SESSION);
  }, [chamber]);

  return (
    <div className="list-container">
      <Filter
        members={members}
        setFilteredMembers={setFilteredMembers}
        chamber={chamber}
        setChamber={setChamber}
        session={session}
        setSession={setSession}
        minimumSession={minimumSession}
      />

      {loading && (
        <div className="loading-container">
          <img src={CongressImg} alt="Loading" className="loading" />
        </div>
      )}

      {!loading &&
        <Pagination
          numberOfMembers={filteredMembers.length}
          currentPage={currentPage}
          setCurrentPage={setCurrentPage}
          membersPerPage={membersPerPage}
          setMembersPerPage={setMembersPerPage}
        />
      }
      {!loading &&
        <Table
          filteredMembers={filteredMembers}
          startIndex={startIndex}
          endIndex={endIndex}
        />
      }
    </div>
  );
};
