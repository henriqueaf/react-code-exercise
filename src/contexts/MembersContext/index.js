import React, { createContext, useEffect, useState, useContext } from 'react';

import { useMembersReducer } from '../../reducers/MembersReducer';
import { setLoading, setMembers, setFilteredMembers, setCurrentPage, setSelectedSession } from '../../reducers/MembersReducer/actions';
import { getMembersCache } from '../../services/membersCache';

const HOUSE_MINIMUM_SESSION = 102;
const SENATE_MINIMUM_SESSION = 80;

const MembersContext = createContext({});

const MembersProvider = ({ children }) => {
  const [state, dispatch] = useMembersReducer();
  const {
    filteredMembers,
    membersPerPage,
    selectedChamber,
    selectedSession,
  } = state;

  const [minimumSession, setMinimumSession] = useState(SENATE_MINIMUM_SESSION);

  const calculateMinimumSession = () => {
    switch (selectedChamber) {
      case 'senate':
        return SENATE_MINIMUM_SESSION;
      case 'house':
        return HOUSE_MINIMUM_SESSION;
    }
  };

  const providerValue = {
    minimumSession,
    state,
    dispatch,
  };

  useEffect(() => {
    dispatch(setLoading(true));

    const url = `https://api.propublica.org/congress/v1/${selectedSession}/${selectedChamber}/members.json`;

    async function membersCache() {
      try {
        const responseMembers = await getMembersCache(url);
        dispatch(setMembers(responseMembers));
        dispatch(setFilteredMembers(responseMembers));
      } catch (error) {
        console.error(error.message);
        dispatch(setLoading(false));
      } finally {
        dispatch(setLoading(false));
        setMinimumSession(calculateMinimumSession());
      }
    }

    membersCache()
  }, [selectedChamber, selectedSession]);

  useEffect(() => {
    if (selectedChamber === 'house' && selectedSession < HOUSE_MINIMUM_SESSION)
      dispatch(setSelectedSession(HOUSE_MINIMUM_SESSION));
  }, [selectedChamber]);

  useEffect(() => {
    dispatch(setCurrentPage(0));
  }, [membersPerPage, filteredMembers]);

  return (
    <MembersContext.Provider value={providerValue}>
      {children}
    </MembersContext.Provider>
  );
};

const useMembersContext = () => useContext(MembersContext);

export { useMembersContext, MembersProvider, MembersContext };
