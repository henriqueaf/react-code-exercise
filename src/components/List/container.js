import React from 'react';

import { MembersProvider } from '../../contexts/MembersContext';
import List from './index';

const Container = () => (
  <MembersProvider>
    <List />
  </MembersProvider>
);

export default Container;
