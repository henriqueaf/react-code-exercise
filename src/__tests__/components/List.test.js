import {useEffect} from 'react';
import {render, fireEvent, screen, cleanup} from '@testing-library/react';
import {MembersContext} from '../../contexts/MembersContext';
import List from '../../components/List';

const defaultContextValues = {
  loading: false,
  setLoading: () => {},
  members: [],
  filteredMembers: [],
  setFilteredMembers: () => {},
  selectedMemberForDetails: null,
  setSelectedMemberForDetails: () => {},
  membersPerPage: 10,
  setMembersPerPage: () => {},
  currentPage: 0,
  setCurrentPage: () => {},
  chamber: 'senate',
  setChamber: () => {},
  session: 115,
  setSession: () => {},
  minimumSession: 80
};

const mount = (props) => {
  const mergedProps = {
    ...defaultContextValues,
    ...props,
  };

  return render(
    <MembersContext.Provider value={mergedProps}>
      <List />
    </MembersContext.Provider>
  );
};

describe('List', () => {
  it('Matches snapshot', () => {
    const {container, getByAltText} = mount();
    expect(container).toMatchSnapshot();
  });

  it('Show loading', () => {
    const {getByAltText} = mount({loading: true});
    const loadingImg = getByAltText('Loading');
    expect(loadingImg).not.toBeNull();
  });
})
