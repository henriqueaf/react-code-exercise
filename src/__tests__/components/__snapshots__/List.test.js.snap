// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Matches snapshot 1`] = `
<div>
  <div
    class="list-container"
  >
    <div
      class="my-3"
    >
      <form
        class=""
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <input
              class="form-control"
              placeholder="Senator/Representative name"
            />
          </div>
          <div
            class="col"
          >
            <select
              class="form-select"
            >
              <option
                value="0"
              >
                Select party
              </option>
              <option
                value="R"
              >
                R
              </option>
              <option
                value="D"
              >
                D
              </option>
            </select>
          </div>
          <div
            class="col"
          >
            <select
              class="form-select"
            >
              <option
                value="0"
              >
                Select gender
              </option>
              <option
                value="M"
              >
                M
              </option>
              <option
                value="F"
              >
                F
              </option>
            </select>
          </div>
          <div
            class="col-sm-1"
          >
            <button
              class="btn btn-primary"
              type="reset"
            >
              Clear
            </button>
          </div>
        </div>
      </form>
      <div
        class="mt-3 row"
      >
        <div
          class="col-sm-2"
        >
          <select
            class="form-select"
          >
            <option
              value="senate"
            >
              senate
            </option>
            <option
              value="house"
            >
              house
            </option>
          </select>
        </div>
        <div
          class="col-sm-2"
        >
          <select
            class="form-select"
          >
            <option
              value="80"
            >
              80
            </option>
            <option
              value="81"
            >
              81
            </option>
            <option
              value="82"
            >
              82
            </option>
            <option
              value="83"
            >
              83
            </option>
            <option
              value="84"
            >
              84
            </option>
            <option
              value="85"
            >
              85
            </option>
            <option
              value="86"
            >
              86
            </option>
            <option
              value="87"
            >
              87
            </option>
            <option
              value="88"
            >
              88
            </option>
            <option
              value="89"
            >
              89
            </option>
            <option
              value="90"
            >
              90
            </option>
            <option
              value="91"
            >
              91
            </option>
            <option
              value="92"
            >
              92
            </option>
            <option
              value="93"
            >
              93
            </option>
            <option
              value="94"
            >
              94
            </option>
            <option
              value="95"
            >
              95
            </option>
            <option
              value="96"
            >
              96
            </option>
            <option
              value="97"
            >
              97
            </option>
            <option
              value="98"
            >
              98
            </option>
            <option
              value="99"
            >
              99
            </option>
            <option
              value="100"
            >
              100
            </option>
            <option
              value="101"
            >
              101
            </option>
            <option
              value="102"
            >
              102
            </option>
            <option
              value="103"
            >
              103
            </option>
            <option
              value="104"
            >
              104
            </option>
            <option
              value="105"
            >
              105
            </option>
            <option
              value="106"
            >
              106
            </option>
            <option
              value="107"
            >
              107
            </option>
            <option
              value="108"
            >
              108
            </option>
            <option
              value="109"
            >
              109
            </option>
            <option
              value="110"
            >
              110
            </option>
            <option
              value="111"
            >
              111
            </option>
            <option
              value="112"
            >
              112
            </option>
            <option
              value="113"
            >
              113
            </option>
            <option
              value="114"
            >
              114
            </option>
            <option
              value="115"
            >
              115
            </option>
            <option
              value="116"
            >
              116
            </option>
            <option
              value="117"
            >
              117
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <ul
          class="pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                «
              </span>
              <span
                class="visually-hidden"
              >
                First
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                ‹
              </span>
              <span
                class="visually-hidden"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                ›
              </span>
              <span
                class="visually-hidden"
              >
                Next
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                »
              </span>
              <span
                class="visually-hidden"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-sm-1"
      >
        <select
          class="form-select"
        >
          <option>
            Select members per page
          </option>
          <option
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="30"
          >
            30
          </option>
        </select>
      </div>
    </div>
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            #
          </th>
          <th>
            Name
          </th>
          <th>
            Twitter
          </th>
          <th>
            Facebook
          </th>
          <th>
            Youtube
          </th>
          <th>
            Party
          </th>
          <th>
            District/State
          </th>
          <th>
            Next Election
          </th>
          <th>
            Details
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
